### ISSUE TITLE
Docker failed to pull image in PR builder

### ISSUE THREAD

**Issue by ihmccreery** [kubernetes/kubernetes#21517]
http://pr-test.k8s.io/21515/kubernetes-pull-build-test-e2e-gce/29272/

```
+++ [0218 15:19:03] Docker builds done
+++ [0218 15:19:03] Pulling and writing Docker image for addon: gcr.io/google_containers/pause:2.0
+++ [0218 15:19:03] Pulling and writing Docker image for addon: gcr.io/google_containers/kube-registry-proxy:0.3
Pulling repository gcr.io/google_containers/kube-registry-proxy
2.0: Pulling from google_containers/pause
8f216013977d: Already exists
a3ed95caeb02: Already exists
Digest: sha256:9ce5316f9752b8347484ab0f6778573af15524124d52b93230b9a0dcc987e73e
Status: Image is up to date for gcr.io/google_containers/pause:2.0
Could not reach any registry endpoint
!!! Error in /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:901
  '"${DOCKER[@]}" pull "${addon_path}"' exited with status 1
Call stack:
  1: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:901 kube::release::write_addon_docker_images_for_server(...)
  2: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:785 kube::release::package_server_tarballs(...)
  3: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:709 kube::release::package_tarballs(...)
  4: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/release.sh:40 main(...)
Exiting with status 1
!!! [0218 15:19:07] unable to pull or write addon image
!!! Error in /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:892
  'kube::release::write_addon_docker_images_for_server "${release_stage}/addons"' exited with status 1
Call stack:
  1: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:892 kube::release::write_addon_docker_images_for_server(...)
  2: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:785 kube::release::package_server_tarballs(...)
  3: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/../build/common.sh:709 kube::release::package_tarballs(...)
  4: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/release.sh:40 main(...)
Exiting with status 1
!!! [0218 15:19:07] previous tarball phase failed
!!! Error in /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/release.sh:712
  'return 1' exited with status 1
Call stack:
  1: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/../build/release.sh:712 main(...)
Exiting with status 1
2016/02/18 15:19:07 e2e.go:307: Error running build-release: exit status 1
2016/02/18 15:19:07 e2e.go:303: Step 'build-release' finished in 4m22.95795267s
2016/02/18 15:19:07 e2e.go:108: Error building. Aborting.
exit status 1
```

**Comment by fejta** [comment]
Any insight into why the docker command failed?

**Comment by ixdy** [comment]
I'm guessing transient failure, but I don't really know.

**Comment by bprashanth** [comment]
```
Status: Image is up to date for gcr.io/google_containers/pause:2.0
Could not reach any registry endpoint
!!! Error in /var/lib/jenkins/workspace/k
```

Sounds similar to https://github.com/kubernetes/kubernetes/issues/20550, which I believe we're just addressing with retries (punting on the deeper issue of diagnosing potential flake in reaching gcr)

**Comment by ixdy** [comment]
Are we still seeing this?

**Comment by bgrant0607** [comment]
See also #13288 (wrt docker hub, not gcr)

**Comment by nikhiljindal** [comment]
This failed again: https://github.com/kubernetes/kubernetes/pull/24104#issuecomment-209715933
```
Pulling repository gcr.io/google_containers/kube-registry-proxy
2.0: Pulling from google_containers/pause
Digest: sha256:9ce5316f9752b8347484ab0f6778573af15524124d52b93230b9a0dcc987e73e
Status: Image is up to date for gcr.io/google_containers/pause:2.0
Could not reach any registry endpoint
!!! Error in /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/gce/../../cluster/../build/../build/common.sh:883
  '"${DOCKER[@]}" pull "${addon_path}"' exited with status 1
```

Not sure why the bug was closed. Was it because it didnt reoccur for a while?

**Comment by fejta** [comment]
Looks like this is using very old code:

``sh
!!! Error in /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/hack/e2e-internal/../../cluster/gce/../../cluster/gce/../../cluster/../build/../build/common.sh:883
  '"${DOCKER[@]}" pull "${addon_path}"' exited with status 1
``

However head shows that we no longer pull if the image already exists:

https://github.com/kubernetes/kubernetes/blame/master/build/common.sh#L884

        if [[ -z $("${DOCKER[@]}" images | awk '{print ($1":"$2)}' | grep "${addon_path}") ]]; then
885	          kube::log::status "Addon image ${addon_path} does not exist, pulling it..."
886	          "${DOCKER[@]}" pull "${addon_path}"
887	        fi

**Comment by fejta** [comment]
@nikhiljindal we fixed this for 1.3 but your PR is for 1.2

**Comment by ixdy** [comment]
Maybe we should cherrypick #23040 to 1.2?
